qiime2/sdk/result.py

```diff
@@ -652,6 +654,101 @@ def data_initializer(destination):
             provenance_capture=provenance_capture)
         return viz
 
+    @classmethod
+    def make_report(cls, template, collection):
+        """Make a report out of existing visualizations and a template.
+
+        Parameters
+        ----------
+        template : Callable[[str, dict], None]
+            A template function which is given the output directory as the
+            first argument and an index of subfigures as the second.
+
+        collection : dict[str, Visualization]
+            The visualizations to make available to the template. It is up
+            to the template to use them, but they will exist in a specialized
+            subfigures directory unique to the report format for Visualization.
+
+        Returns
+        -------
+        Visualization
+            The newly created report as a Visualization. It will have the
+            format of "report".
+
+        """
+        provenance_capture = archive.ReportProvenanceCapture()
+        to_reindex = {}
+
+        def data_initializer(destination):
+            index = {}
+            subfigures_dir = os.path.join(destination, 'subfigures')
+            for key, viz in collection.items():
+                viz_uuid = str(viz.uuid)
+                provenance_capture.add_input(key, viz)
+                index[key] = {
+                    "name": key,
+                    "index": f'subfigures/{viz_uuid}/index.html',
+                }
+                subfigure_root = os.path.join(subfigures_dir, viz_uuid)
+                if not os.path.exists(subfigure_root):
+                    shutil.copytree(viz._archiver.data_dir, subfigure_root)
+
+                # if not for reports of reports, this would have been the end
+                # of the data_initializer, however we don't want nested
+                # reports to have an arbitrarily long path as this will break
+                # things. Also we get de-duplication for free if we hoist
+                # sub-subfgures into just subfigures of the outermost report
+                if viz.format is report:
+                    # reports which were providfed as part of the collection
+                    # will have sub-figures moved and index.json re-written
+                    # to a new flattened path
+                    to_reindex[viz_uuid] = set()
+                    child_figs = os.path.join(subfigure_root, 'subfigures')
+                    with open(os.path.join(child_figs, 'index.json')) as fh:
+                        inner_index = json.load(fh)
+                        index[key]['children'] = inner_index
+                        to_reindex[viz_uuid] = \
```

Comment:
I'd probably leave a comment explaining this a bit, in particular, probably explaining what I expect the path I'm splitting and grabbing the -2 element from to look like.